cloudsmith/data_source_package_download_test.go

```diff
@@ -0,0 +1,196 @@
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+	"testing"
+	"time"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
+)
+
+func TestAccPackageDownload_data(t *testing.T) {
+	t.Parallel()
+
+	namespace := os.Getenv("CLOUDSMITH_NAMESPACE")
+	apiKey := os.Getenv("CLOUDSMITH_API_KEY")
+	userName := "token"
+	repoName := "terraform-download-test"
+	repoConfig := testAccDataSourcePackageDownload_repositoryConfig(repoName, namespace)
+	fileName := "hello.txt"
+	packageName := "hello"
+	packageVersion := "1.0"
+	destinationPath := "."
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:  func() { testAccPreCheck(t) },
+		Providers: testAccProviders,
+		Steps: []resource.TestStep{
+			{
+				Config: repoConfig,
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("cloudsmith_repository.test", "name", repoName),
+					resource.TestCheckResourceAttr("cloudsmith_repository.test", "namespace", namespace),
+				),
+			},
+			{
+				PreConfig: func() {
+					createHelloWorldFile(fileName)
+					uploadPackageToCloudsmith(apiKey, namespace, repoName, fileName, packageName, packageVersion, userName)
+				},
+				Config: testAccPackageDownloadData(namespace, repoName, packageName, packageVersion, destinationPath),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("data.cloudsmith_package_download.test", "namespace", namespace),
+					resource.TestCheckResourceAttr("data.cloudsmith_package_download.test", "repository", repoName),
+					resource.TestCheckResourceAttr("data.cloudsmith_package_download.test", "package_name", packageName),
+					resource.TestCheckResourceAttr("data.cloudsmith_package_download.test", "package_version", packageVersion),
+					resource.TestCheckResourceAttr("data.cloudsmith_package_download.test", "destination_path", destinationPath),
+					checkDownloadedFile(namespace, repoName, packageName, packageVersion, destinationPath, "Hello world"),
+				),
+			},
+		},
+	})
+}
+
+func testAccDataSourcePackageDownload_repositoryConfig(repoName string, namespace string) string {
+	return fmt.Sprintf(`
+resource "cloudsmith_repository" "test" {
+	name      = "%s"
+	namespace = "%s"
+}
+`, repoName, namespace)
+}
+
+func createHelloWorldFile(filename string) {
+	content := []byte("Hello world")
+	err := os.WriteFile(filename, content, 0644)
+	if err != nil {
+		panic(err)
+	}
+}
+
+func uploadPackageToCloudsmith(apiKey, namespace, repoName, fileName, packageName, packageVersion string, userName string) {
+	client := &http.Client{}
+
+	// Calculate the SHA-256 checksum of the file
+	fileData, err := os.ReadFile(fileName)
+	if err != nil {
+		panic(err)
+	}
+	sha256Checksum := fmt.Sprintf("%x", sha256.Sum256(fileData))
+
+	// Step 1: PUT (upload) the file
+	uploadURL := fmt.Sprintf("https://upload.cloudsmith.io/%s/%s/%s", namespace, repoName, fileName)
```

Comment:
can we avoid this hardcoded URL here? otherwise the provider won't work against stg/dev/local envs